### DISCUSSION TITLE
Why Isn't `dotMultiply` Working With Scalars?

### DISCUSSION THREAD

**Discussion by fweth** [josdejong/mathjs#2730] [Ideas]
Hello, I think it would be useful if something like `math.dotMultiply([1, -1, 1], [[1, 2], [3, 4], [5, 6]])` would return `[[1, 2], [-3, -4], [5, 6]]` as currently this results in a dimension mismatch.

The logic could be, check if the shape of the second argument can be obtained from the shape of the first agument by replacing all scalars with something that can be multiplied with a scalar (e.g. a scalar, a vector, a matrix), in that case, do scalar multiplication.

**Comment by josdejong** [comment]
Thanks for sharing your idea. I understand what you mean but it feels complex and hard to explain to me. Right now, `dotMultiply` simply requires two matrices having the same size and multiplies them element-wise.

**Reply by fweth** [reply]
I see, thanks for replying!  Julia has [broadcasting](https://docs.julialang.org/en/v1/manual/arrays/#Broadcasting), which allows something like `[1, 2, 3] .* [[1, 2], [3, 4], [5, 6]]` resulting in `[[1, 2], [6, 8], [15, 18]]` but I never really understood the exact specifications of that, so I guess you are right, there is a lot of hidden complexity.

**Reply by josdejong** [reply]
>  but I never really understood the exact specifications, so I guess you are right

😂  

If there is a clear use case I'm always in to discuss it. Maybe not implement it with the `.*` operator but a specific named function like `multiplyColumns` and `multiplyRows` or so.

**Reply by fweth** [reply]
I think my problem is that I often want to use one library which covers all of my usecases, but here it would be probably wise to use something like Ramda and do `R.zipWith(math.multiply, [1, 2, 3], [[1, 2], [3, 4], [5, 6]])`...

**Reply by josdejong** [reply]
ah, that is a smart one 👍

**Comment by josdejong** [comment]
I just came across an old discussion about the same idea: #2753

**Comment by gwhitney** [comment]
Or, the diagonal matrix with entries 1, -1, 1 along the diagonal (I think mathjs has a function to produce that) times [[1,2],[3,4],[5,6]] also gives what you wanted in your original question.

**Reply by josdejong** [reply]
💡👍
